### ISSUE TITLE
Wrap default HTML5 backend

### ISSUE THREAD

**Issue by choffmeister** [yahoo/react-dnd-touch-backend#35]
Hi,

I just got an idea I wanted to discuss here: As I see in the source code this DND backend uses `mousedown/up/move` to also work in desktop machines, but this has some edge cases. I ask myself, if it would be possible to wrap the original HTML5 backend (like some kind of higher order backend) that handles the touch case but also passes the native DND events to HTML5 backend, so that one can have both at a time.

Did someone already try this? Or does this not work with the react-dnd backend interface?

**Comment by longlho** [comment]
It's certainly possible, the complexity is in event propagation differences between touch events & mouse events, along w/ native interpretation of tap, long tap (bring up context menu) and such. DnD are targeted APIs by browsers so it's easier to assume the intention of the action

**Comment by shotaK** [comment]
What about conditionally loading different backends for different devices. e.g we can check whether device is touch and if it is we can load touch backend and so on. Also doc says that mouse events support is buggy, does anyone know what this exactly means? is it dangerous to use it for production, which is supposed to work on both desktop and touch devices?

**Comment by felipeleusin** [comment]
Hi,
I just used: ('ontouchstart' in document.documentElement) ? TouchBackend : HTML5Backend and worked but I need to test in more devices

**Comment by longlho** [comment]
@shotaK the mouse support is buggy in certain edge cases. For ex:

When you "long" `touch` an `anchor` tag in touch device, it'll bring up the context menu to bookmark, etc... Now since mouse listens on `mousedown`, it'll conflict w/ this behavior since you'd have to long `touch` it to drag it somewhere. If you don't use `anchor` tag to drag, it should be fine.

In general, long `touch` depending on which element it is has different behavior. Mouse propagation also works differently as well, so that's definitely an issue.

**Comment by shotaK** [comment]
https://github.com/longlho thanks for the explanation.

I also started loading it conditionally and so far it works great.

```
import Modernizr from 'browsernizr';

export default DragDropContext(Modernizr.touchevents ? TouchBackend({ enableMouseEvents: true }) : HTML5Backend)(SelectedTilesList);

```

**Comment by clessg** [comment]
Unfortunately, [this article](http://www.stucox.com/blog/you-cant-detect-a-touchscreen/) still seems to be relevant.

Something like `Modernizr.touchevents` works most of the time, at least. Unfortunately, hybrid devices are kind of screwed, as mouse-based dragging will no longer function.

**Comment by koutsenko** [comment]
I don't want to use HTML5 backend because there is no way to disable dragging ghost transparence. Is it possible to create a topic where we can discuss all mouse-related **bugs** and **workarounds** when **enableMouseEvents** was set to **true**?

**Comment by choffmeister** [comment]
You can disable it. You can hide native drag preview alltogether and build a custom drag preview.

**Comment by koutsenko** [comment]
@choffmeister , yes, but that custom preview will have opacity ~0.5 and I can't override that specs-related behaviour...

**Comment by choffmeister** [comment]
No. You completely hide it. so u have no html5 drag preview. And then u create a new DOM node which u control with javascript do float next to the mouse cursor.

**Comment by choffmeister** [comment]
@koutsenko See https://react-dnd.github.io/react-dnd/docs-drag-layer.html for more information

**Comment by koutsenko** [comment]
@choffmeister thanks, okay, what if I use hybrid touch device (laptop with touchscreen). What backend i should use when page loads - react-dnd-html5-backend for mouse\touchpad or react-dnd-touch-backend for touchscreen? They can't work together, and making choice will disable one input type. Seems that second backend can handle both input types. But... there is a notice about mouse-related bugs.

**Comment by koutsenko** [comment]
I'm thinking about third-party libs like [that](https://medium.com/@michael.sotkin/creating-easy-drag-and-drop-components-with-interact-js-in-react-3060784bf843#.chtqc7ahy) , of course, I'll replace findDOMNode by refs, but it's hacky solution also - I must inject into lifecycles of all applications components which take an participance in drag-n-drop operations

**Comment by JoaoMosmann** [comment]
It's a bit late, but it seems this guy is doing a good job related to this issue

https://github.com/LouisBrunner/react-dnd-multi-backend
